src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,304 @@
+            {
+                return;
+            }
+
+            IMethodSymbol? preferredWriteAsyncMethod = writeAsyncMethodGroup.FirstOrDefault(x =>
+                x.Parameters.Count() == 2 && x.Parameters[0].Type is INamedTypeSymbol type && type.ConstructedFrom.Equals(readOnlyMemoryType));
+            if (preferredWriteAsyncMethod == null)
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+
+                IOperation parentOperation = invocation.Parent;
+                if (parentOperation == null)
+                {
+                    return;
+                }
+                // Only accept these two cases:
+                // - await {WriteAsync()|ReadAsync()}
+                // - await {WriteAsync()|ReadAsync()}.ConfigureAwait()
+                if (parentOperation.Kind == OperationKind.Await ||
+                        (parentOperation.Parent != null && parentOperation.Parent.Kind == OperationKind.Await &&
+                         parentOperation.Parent is IAwaitOperation awaitOperation && awaitOperation.Operation is IInvocationOperation grandparentInvocation &&
+                            (grandparentInvocation.TargetMethod.OriginalDefinition.Equals(configureAwaitMethod) ||
+                             grandparentInvocation.TargetMethod.OriginalDefinition.Equals(genericConfigureAwaitMethod))))
+                {
+                    // Verify if the current method's type is or inherits from Stream
+                    if (!IsDefinedBy(invocation.TargetMethod, streamType, out IMethodSymbol method))
+                    {
+                        return;
+                    }
+
+                    DiagnosticDescriptor rule;
+                    string ruleMessageMethod;
+                    string ruleMessagePreferredMethod;
+
+                    // Verify if the method is an undesired Async overload
+                    if (method.Equals(undesiredReadAsyncMethod) || method.Equals(undesiredReadAsyncMethodWithCancellationToken))
+                    {
+                        rule = PreferStreamReadAsyncMemoryOverloadsRule;
+                        ruleMessageMethod = undesiredReadAsyncMethod.Name;
+                        ruleMessagePreferredMethod = preferredReadAsyncMethod.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat);
+                    }
+                    else if (method.Equals(undesiredWriteAsyncMethod) || method.Equals(undesiredWriteAsyncMethodWithCancellationToken))
+                    {
+                        rule = PreferStreamWriteAsyncMemoryOverloadsRule;
+                        ruleMessageMethod = undesiredWriteAsyncMethod.Name;
+                        ruleMessagePreferredMethod = preferredWriteAsyncMethod.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat);
+                    }
```

Comment:
Comparing the target method symbol with one of the undesired methods is the cheapest check and should be the first one to be done. i.e. all the earlier checks related to the call being in context of await/ConfigureAwait/IsDefinedBy should be done after these cheaper checks

Reply:
I can do that except for IsDefinedBy: That's the method I need to retrieve the actual IMethodSymbol to compare against.

Reply:
Nevermind, found a way.